### PR TITLE
For `Optional[T] or T`, infer the type `T` instead of `Optional[T]`.

### LINKED ISSUE #1733: Type inference of `or` and `Optional[x]`
When using `--strict-optional`, the type of `a` is incorrectly `Union[None, str]`:

```
from typing import Optional

def f(x: Optional[str]) -> None:
    a = x or 'default'
    reveal_type(a)  # should be str
```

#1698 seems related.

### LINKED ISSUE #1817: Improve type inference for "Optional[T] or T" expression
Consider this:
```python
def f(a: Optional[str]) -> str:
    return a or 'default'
```
This should type-check under `--strict-optional`, but currently the type of `a or 'default'` is inferred as `Optional[str]`. It should realize that if a is `None` the right branch is always taken.

### PR DESCRIPTION
**Pull request by gvanrossum**
Fixes #1817, fixes #1733.

### CODE REVIEW

**Comment by gvanrossum** [comment]
Is there an equivalent example for `and`? I can't think of any.

FWIW this is kind of blocking further explorations of `--strict-optional` on mypy itself.

File: test-data/unit/check-optional.test [hunk 1]
```diff
@@ -175,3 +175,33 @@ def f(x: None) -> str: pass
 def f(x: int) -> int: pass
 reveal_type(f(None))  # E: Revealed type is 'builtins.str'
 reveal_type(f(0))  # E: Revealed type is 'builtins.int'
+
+[case testOptionalTypeOrTypePlain]
+from typing import Optional
+def f(a: Optional[int]) -> int:
```
Comment (JukkaL) [inline test-data/unit/check-optional.test]: Maybe test `or` with both operands optional?

Comment (JukkaL) [inline test-data/unit/check-optional.test]: Also, could test left operand with a complex union type such as `Union[int, str, None]`.

Comment (gvanrossum) [inline test-data/unit/check-optional.test]: Done; I've also added special-casing for `None or T` (see tests for what I
mean).

**Comment by ddfisher** [comment]
I feel like we shouldn't need this special casing and this should just fall out of the `find_isinstance_check` like everything else.  I'm a bit surprised that that isn't the case already -- going to briefly investigate why that is.

**Comment by ddfisher** [comment]
Looks like that was actually pretty straightforward to do.  Let's consider #1828 instead of this?

**Comment by gvanrossum** [comment]
Closing in favor of #1828.

**Comment by gvanrossum** [comment]
Cool. Maybe someday I'll understand how this works -- the name (find_isinstance_check()) and the resulting variables (right_map, left_map -- why the transposition?) were all kind of confusing.

**Comment by ddfisher** [comment]
Agree it's all pretty confusing.  The function might be be better named `find_conditional_type_information` or similar.  The transposition is because for `and` we care about the type of the left hand expression if it's falsey but for `or` we care about the value of the left hand expression when it's truthy (and `find_isinstance_check` returns a (conditional type map if true, conditional type map if false) tuple).
